enhancements/api-review/custom-tags-aws.md

```diff
@@ -75,10 +75,70 @@ will apply these tags to all AWS resources they create.
 
 userTags that are specified in the infrastructure resource will merge with userTags specified in an individual component. In the case where a userTag is specified in the infrastructure resource and there is a tag with the same key specified in an individual component, the value from the individual component will have precedence and be used.
 
-The userTags field is intended to be set at install time and is considered immutable. Components that respect this field must only ever add tags that they retrieve from this field to cloud resources, they must never remove tags from the existing underlying cloud resource even if the tags are removed from this field(despite it being immutable).
+Users can update the `resourceTags` by editing `.spec.aws` of the `infrastructure.config.openshift.io` type. New addition of tags are always appended. Any update in the existing tags, which are added by installer or  by edit in `resourceTags`, will replace the previous tag value.
+
+`.status.platformStatus.aws.resourceTags` reflects the present set of userTags. In case when the userTags are created by installer or newly added in infrastructure resource is updated on the individual component directly using external tools, the value from infrastructure resource will have the precedence.
+
+The precedence helps to maintain creator/updator tool (in-case of external tool usage) remains same for user-defined tags which are created correspondingly.
+
+The userTags field is intended to be set at install time and updatable (not allowed to delete). Components that respect this field must only ever add tags that they retrieve from this field to cloud resources, they must never remove tags from the existing underlying cloud resource even if the tags are removed from this field.
```

Comment:
We can keep clarity of Desired (spec) and Observed (status) and still restrict unsupported values - with [dynamic admission control](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/).

Reply:
We can use a dynamic admission control for this, though at the moment, I don't think we have anything set up to observe the infrastructure object.

If that's the route you want to take here, I would consult the API team for their suggestion as you will need to create and manage a validating and/or mutating webhook which comes with some overhead

Reply:
This is great point, thanks.

The idea is to have a "observed config" way to be aware of how Openshift deals with the tags. 

> `.status.platformStatus.aws.resourceTags` reflects the present set of userTags. In case when the userTags are created by installer or newly added in infrastructure resource is updated on the individual component directly using external tools, the value from infrastructure resource will have the precedence.

Also, the precedence of user tag updates can be maintained.

> The precedence helps to maintain creator/updator tool (in-case of external tool usage) remains same for user-defined tags which are created correspondingly.

Reply:
> feel like we probably will want to have some way to deny removal of tags from the infrastructure spec.

just my 2cents here. I think it's totally fine from a customer PoV to remove tags from the infra spec to set these tags as "unmanaged" by OpenShift. Imagine switching to a 3rd party component to manage tags like that one: https://github.com/mtougeron/k8s-aws-ebs-tagger

Would be bad if a customer would need to reinstall the entire cluster to get rid of the tag management.

What we continue to NOT do, is try to be smart about deleting existing tags on resources or detecting what we created vs. what was already there and such.

Reply:
> Would be bad if a customer would need to reinstall the entire cluster to get rid of the tag management.

This is a pretty good argument, I'd say this needs writing out in the enhancement to explain that if that's the route we want to take. The existing language saying not to delete the tags will need to be updated/clarified

Reply:
> is how we will deal with the 25 tag limit from users
The reality is even worse. Since S3 buckets can have at most 10 tags, and OpenShift is already using 2 of those, the user can only add 8 more tags.